### DISCUSSION TITLE
Support Generics

### DISCUSSION THREAD

**Discussion by pmartelletti** [filamentphp/filament#8214] [Ideas]
Many of the Liveweire Components Filament extends when creating a new resource make use of the "Model" Laravel class. This leads that, when running tools like PHPStan it complains about a method or property not existing in the Model object. Which, fair enough, is a truth assertion.

To avoid these errors, one could make use of Generics. For example, the `ViewRecord` class could be slightly changed to look like:

```php
/**
 * @property Form $form
 * @template T of Model
 */
class ViewRecord extends Page implements HasInfolists
{
    use Concerns\HasRelationManagers;
    /** @use InteractsWithRecord<T> */
    use Concerns\InteractsWithRecord;
}
```

At the same time, one would need to change the `InteractsWithRecord` trait to look like so:

```php
/**
 * @template T of Model
 */
trait InteractsWithRecord
{
/**
     * @return T
     */
    public function getRecord() 
    {
        return $this->record;
    }
```

Then, in the user's codebase, one would define the ViewRecordPage as follows (it can be done using the stub automatically):

```php
/**
 * @extends ViewRecord<StudentLesson>
 */
class ViewStudentLesson extends ViewRecord
{}
```

This means that PHPStan will know that, whenever you call `$this->getRecord()` in this component, one has an instance of `StudentLesson`, and not a generic `Model`.

Thoughts on this approach? I'm conscious Filament has just had a big refactor, so maybe not supporting generics was a conscious decision. 

Thanks,

**Comment by danharrin** [comment]
I'm just concerned about maintenance burden, if it isnt too bad then go ahead and PR

**Reply by pmartelletti** [reply]
To be honest, I'm not 100% sure how much work would be involved. At the moment I've tried it just with the classes PHPStan was complaining about for my particular use case (i.e., my app).  So it might be that it will become a PITA to maintain - but if people think it's worth it (i.e., it adds value), then I might spend a couple of days in the PR.

**Reply by danharrin** [reply]
I usually just override `getRecord()` and return the correct model type

**Reply by pmartelletti** [reply]
Right, that works, yes.

I was trying to override it by doing:

```php
/**
     * @return StudentLesson
     */
    public function getRecord() : StudentLesson
    {
        return $this->record;
    }
```

and it was complaining because the return type was not the same as the parent method. Changing it to:

```php
/**
     * @return StudentLesson
     */
    public function getRecord() : Model
    {
        return $this->record;
    }
```

works, PHPStan & PHPStorm are now happy. I've lost a couple of hours trying with the code changes above - I feel a bit stupid now 😢 

I'll close this discussion then, if that's ok?

**Reply by danharrin** [reply]
Sounds good!
